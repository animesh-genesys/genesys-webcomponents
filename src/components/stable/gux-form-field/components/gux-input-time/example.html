<form onchange="notify(event)" oninput="notify(event)">
  <h1>gux-form-field (Time)</h1>

  <h3>Default</h3>
  <gux-form-field label-position="screenreader">
    <input name="time" type="time" slot="input" />
    <label slot="label">This will be read by a screenreader</label>
  </gux-form-field>

  <h3>With Label</h3>
  <gux-form-field label-position="above">
    <input name="time" type="time" slot="input" />
    <label slot="label">Select Time</label>
  </gux-form-field>

  <h3>Interval: 30 Minutes</h3>
  <gux-form-field interval="30" label-position="screenreader">
    <input name="time" type="time" slot="input" />
    <label slot="label">This will be read by a screenreader</label>
  </gux-form-field>

  <h3>Disabled</h3>
  <gux-form-field label-position="screenreader">
    <input name="time" type="time" slot="input" disabled />
    <label slot="label">This will be read by a screenreader</label>
  </gux-form-field>

  <h3>Error</h3>
  <gux-form-field label-position="screenreader">
    <input name="time" type="time" slot="input" />
    <label slot="label">This will be read by a screenreader</label>
    <span slot="error">Please provide a valid time</span>
  </gux-form-field>
</form>
