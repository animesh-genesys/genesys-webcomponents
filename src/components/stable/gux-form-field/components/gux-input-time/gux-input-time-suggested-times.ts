// 12 Hour
export const Gux15MinuteInterval = [
  '12:00',
  '12:15',
  '12:30',
  '12:45',
  '1:00',
  '1:15',
  '1:30',
  '1:45',
  '2:00',
  '2:15',
  '2:30',
  '2:45',
  '3:00',
  '3:15',
  '3:30',
  '3:45',
  '4:00',
  '4:15',
  '4:30',
  '4:45',
  '5:00',
  '5:15',
  '5:30',
  '5:45',
  '6:00',
  '6:15',
  '6:30',
  '6:45',
  '7:00',
  '7:15',
  '7:30',
  '7:45',
  '8:00',
  '8:15',
  '8:30',
  '8:45',
  '9:00',
  '9:15',
  '9:30',
  '9:45',
  '10:00',
  '10:15',
  '10:30',
  '10:45',
  '11:00',
  '11:15',
  '11:30',
  '11:45'
];

export const Gux30MinuteInterval = [
  '12:00',
  '12:30',
  '1:00',
  '1:30',
  '2:00',
  '2:30',
  '3:00',
  '3:30',
  '4:00',
  '4:30',
  '5:00',
  '5:30',
  '6:00',
  '6:30',
  '7:00',
  '7:30',
  '8:00',
  '8:30',
  '9:00',
  '9:30',
  '10:00',
  '10:30',
  '11:00',
  '11:30'
];

// 24 Hour
