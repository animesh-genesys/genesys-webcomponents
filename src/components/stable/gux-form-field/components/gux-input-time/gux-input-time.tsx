import {
  Component,
  Element,
  Event,
  h,
  JSX,
  Listen,
  Prop,
  State
} from '@stencil/core';
import { buildI18nForComponent, GetI18nValue } from '../../../../../i18n';
import componentResources from './i18n/en.json';
import { onDisabledChange } from '../../../../../utils/dom/on-attribute-change';
import { whenEventIsFrom } from '../../../../../utils/dom/when-event-is-from';
import { OnClickOutside } from '../../../../../utils/decorator/on-click-outside';
import {
  Gux15MinuteInterval,
  Gux30MinuteInterval
} from './gux-input-time-suggested-times';

const MAX_HOURS: string = '12';
const MAX_MINUTES: string = '59';
const MIN_HOURS: string = '1';
const MIN_MINUTES: string = '00';

/**
 * @slot input - Required slot for input[type="time"]
 */
@Component({
  styleUrl: 'gux-input-time.less',
  tag: 'gux-input-time'
})
export class GuxInputTime {
  private i18n: GetI18nValue;
  private input: HTMLInputElement;
  private disabledObserver: MutationObserver;

  @Element()
  private root: HTMLElement;

  /**
   * AM/PM toggle value
   */
  @Prop({ mutable: true })
  toggleAmPMValue: string = '';

  /**
   * Suggested times dropdown list state - open/closed
   */
  @Prop({ mutable: true })
  opened: boolean = false;

  /**
   * Indicate the dropdown input value
   */
  @Prop({ mutable: true })
  value: string = '';

  /**
   * Hours input value
   */
  @Prop({ mutable: true })
  hoursValue: string = '12';

  /**
   * Minutes input value
   */
  @Prop({ mutable: true })
  minutesValue: string = '00';

  /**
   * Time interval between suggested times in dropdown list
   */
  @Prop()
  interval: string = '15';

  @State()
  private disabled: boolean;

  @State()
  clockActive: boolean = false;

  async componentWillLoad(): Promise<void> {
    this.i18n = await buildI18nForComponent(this.root, componentResources);
    this.input = this.root.querySelector('input[slot="input"]');

    this.disabled = this.input.disabled;
    this.toggleAmPMValue = this.i18n('am');

    this.disabledObserver = onDisabledChange(
      this.input,
      (disabled: boolean) => {
        this.disabled = disabled;
      }
    );
  }

  @Listen('keydown', { passive: false })
  onKeyDown(event: KeyboardEvent) {
    const selectionOptions = this.getTimeOptions();
    const focusIndex = this.getFocusIndex(selectionOptions);

    switch (event.key) {
      case 'ArrowDown':
        event.preventDefault();

        if (this.clockActive) {
          if (focusIndex < selectionOptions.length - 1) {
            selectionOptions[focusIndex + 1].focus();
          }
        } else {
          this.increment(-1, event);
        }
        break;
      case 'ArrowUp':
        event.preventDefault();

        if (this.clockActive) {
          if (focusIndex > 0) {
            selectionOptions[focusIndex - 1].focus();
          }
        } else {
          this.increment(1, event);
        }
        break;
      case 'Tab':
        if (this.clockActive) {
          this.opened = false;
        }
        break;
      case 'Enter':
      case 'Space':
        if (this.clockActive) {
          this.optionSelectedHandler(event);
        }
        break;
      default:
        break;
    }
  }

  @OnClickOutside({ triggerEvents: 'mousedown' })
  onClickOutside(e: MouseEvent) {
    if (!e.relatedTarget || !this.root.contains(e.relatedTarget as Node)) {
      this.opened = false;
      this.clockActive = false;
    }
  }

  private getFocusIndex(selectionOptions: HTMLGuxOptionElement[]): number {
    return selectionOptions.findIndex(option => {
      return option.matches(':focus');
    });
  }

  disconnectedCallback(): void {
    this.disabledObserver.disconnect();
  }

  /**
   * For clock button open/close of dropdown list
   */
  private toggle() {
    if (this.opened === false) {
      this.clockActive = true;
    }
    this.opened = !this.opened;
  }

  setValue(value: string) {
    this.value = value;
    this.opened = false;
  }

  /**
   * Create list of suggested times
   */
  suggestedTimesList() {
    const arrListItems = [];
    if (this.interval != '15') {
      Gux30MinuteInterval.forEach(time => {
        arrListItems.push(
          <gux-option class="gux-time-option" value={time}>
            {time}
          </gux-option>
        );
      });
    } else {
      Gux15MinuteInterval.forEach(time => {
        arrListItems.push(
          <gux-option class="gux-time-option" value={time}>
            {time}
          </gux-option>
        );
      });
    }

    return (
      <div class="gux-time-options-list">{arrListItems}</div>
    ) as JSX.Element;
  }

  /**
   * Get time options for suggested times
   */
  private getTimeOptions(): HTMLGuxOptionElement[] {
    const result: HTMLGuxOptionElement[] = [];
    const options: HTMLElement = this.root.getElementsByClassName(
      'gux-time-options'
    )[0] as HTMLElement;

    if (!options) {
      return [];
    }

    const childrenElements = Array.from(options.children[0].children);
    childrenElements.forEach(child => {
      if (child.matches('gux-option')) {
        result.push(child as HTMLGuxOptionElement);
      }
    });

    return result;
  }

  /**
   * Set hours and minutes value when suggested option is selected
   * @param e - option selected event
   */
  private optionSelectedHandler(e: Event) {
    whenEventIsFrom('gux-option', e, elem => {
      const option = elem as HTMLGuxOptionElement;
      const selectionOptions = this.getTimeOptions();

      selectionOptions.forEach(selectionOption => {
        if (selectionOption === option) {
          this.setValue(selectionOption.value);
        }
      });
    });

    this.getHoursMinutesNewValue();
    this.clockActive = false;
  }

  getHoursMinutesNewValue() {
    if (this.value != '') {
      const splitValue = this.value.split(':');
      this.hoursValue = splitValue[0];
      this.minutesValue = splitValue[1];
    }
  }

  /**
   * Increment (or decrement) hours/minutes input value when using keyboard up/down keys
   * @param delta - number to increment or decrement value by
   * @param e - mouse click event
   */
  increment(delta: number, e: Event) {
    const target = e.target as HTMLGuxInputTimeElement;
    const targetInput = target.className;
    const targetValue = target.value;

    if (targetInput === 'gux-input-time-hours') {
      if (delta === 1) {
        if (targetValue === MAX_HOURS) {
          this.hoursValue = MIN_HOURS;
        } else {
          this.hoursValue = (parseInt(targetValue) + 1).toString();
        }
      }

      if (delta === -1) {
        if (targetValue === MIN_HOURS) {
          this.hoursValue = MAX_HOURS;
        } else {
          this.hoursValue = (parseInt(targetValue) - 1).toString();
        }
      }
    }

    if (targetInput === 'gux-input-time-minutes') {
      if (delta === 1) {
        if (targetValue === MAX_MINUTES) {
          this.minutesValue = MIN_MINUTES;
        } else {
          this.minutesValue = (parseInt(targetValue) + 1).toString();
        }
      }

      if (delta === -1) {
        if (targetValue === MIN_MINUTES) {
          this.minutesValue = MAX_MINUTES;
        } else {
          this.minutesValue = (parseInt(targetValue) - 1).toString();
        }
      }

      if (this.minutesValue.length < 2) {
        this.minutesValue = '0' + this.minutesValue;
      }
    }
  }

  onHoursChanged() {
    this.input = this.root.querySelector('.gux-input-time-hours');
    const hoursValue = this.input.value.replace(/\s/g, '');

    if (
      parseInt(hoursValue) <= parseInt(MAX_HOURS) &&
      parseInt(hoursValue) >= parseInt(MIN_HOURS)
    ) {
      this.hoursValue = hoursValue;
    } else if (
      parseInt(hoursValue) > parseInt(MAX_HOURS) ||
      parseInt(hoursValue) < parseInt(MIN_HOURS)
    ) {
      this.hoursValue = MAX_HOURS;
    } else if (hoursValue.length < 1) {
      // 12 Hour
      this.hoursValue = MAX_HOURS;
    } else {
      this.hoursValue = MAX_HOURS;
    }

    this.input.value = this.hoursValue;
  }

  onMinutesChanged() {
    this.input = this.root.querySelector('.gux-input-time-minutes');
    const minutesValue = this.input.value.replace(/\s/g, '');

    if (
      parseInt(minutesValue) > parseInt(MAX_MINUTES) ||
      parseInt(minutesValue) < parseInt(MIN_MINUTES)
    ) {
      this.minutesValue = MIN_MINUTES;
    } else if (minutesValue.length < 1) {
      this.minutesValue = MIN_MINUTES;
    } else if (minutesValue.length < 2) {
      this.minutesValue = '0' + minutesValue;
    } else {
      this.minutesValue = MIN_MINUTES;
    }

    this.input.value = this.minutesValue;
  }

  onAmPmSelectorClicked() {
    if (this.toggleAmPMValue === this.i18n('am')) {
      this.toggleAmPMValue = this.i18n('pm');
    } else {
      this.toggleAmPMValue = this.i18n('am');
    }
  }

  renderAmPmSelector(): JSX.Element {
    return (
      <button
        class="gux-input-time-am-pm-selector"
        type="button"
        disabled={this.disabled}
        aria-label={this.i18n('toggleAmPM', { amOrPm: this.toggleAmPMValue })}
        onClick={() => this.onAmPmSelectorClicked()}
      >
        {this.toggleAmPMValue}
      </button>
    ) as JSX.Element;
  }

  renderClockButton(): JSX.Element {
    return (
      <button
        class={{
          'gux-clock-button': true,
          'gux-active': this.clockActive
        }}
        type="button"
        disabled={this.disabled}
        aria-label={this.i18n('clockButton')}
        onClick={this.toggle.bind(this)}
      >
        <gux-icon decorative icon-name="clock-outline"></gux-icon>
      </button>
    ) as JSX.Element;
  }

  render(): JSX.Element {
    return (
      <div
        class={{
          'gux-input-time': true,
          'gux-disabled': this.disabled
        }}
        aria-disabled={this.disabled.toString()}
        ref={(el: HTMLInputElement) => (this.input = el)}
      >
        <div class="gux-input-time-field">
          <div class="gux-input-time-field-input">
            <div
              class={{
                'gux-input-time-field-text-input': true
              }}
            >
              <input
                class="gux-input-time-hours"
                type="number"
                min={MIN_HOURS}
                max={MAX_HOURS}
                disabled={this.disabled}
                value={this.hoursValue}
                onChange={this.onHoursChanged.bind(this)}
                aria-label={this.i18n('hoursInput')}
              />
              <span
                class={{
                  'gux-time-separator': true,
                  'gux-disabled': this.disabled
                }}
              >
                {this.i18n('colon')}
              </span>
              <input
                class="gux-input-time-minutes"
                type="number"
                min={MIN_MINUTES}
                max={MAX_MINUTES}
                disabled={this.disabled}
                value={this.minutesValue}
                onChange={this.onMinutesChanged.bind(this)}
                aria-label={this.i18n('minutesInput')}
              />
              {this.renderAmPmSelector()}
              {this.renderClockButton()}
            </div>
          </div>
        </div>
        <div
          class={{
            'gux-time-options': true,
            'gux-opened': this.opened,
            'gux-disabled': this.disabled
          }}
          aria-label="time options opened"
          onClick={this.optionSelectedHandler.bind(this)}
        >
          {this.suggestedTimesList()}
        </div>
      </div>
    ) as JSX.Element;
  }
}
