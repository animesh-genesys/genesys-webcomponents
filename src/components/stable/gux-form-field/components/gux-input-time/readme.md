# gux-input-time



<!-- Auto Generated Below -->


## Properties

| Property          | Attribute             | Description                                            | Type      | Default |
| ----------------- | --------------------- | ------------------------------------------------------ | --------- | ------- |
| `hoursValue`      | `hours-value`         | Hours input value                                      | `string`  | `'12'`  |
| `interval`        | `interval`            | Time interval between suggested times in dropdown list | `string`  | `'15'`  |
| `minutesValue`    | `minutes-value`       | Minutes input value                                    | `string`  | `'00'`  |
| `opened`          | `opened`              | Suggested times dropdown list state - open/closed      | `boolean` | `false` |
| `toggleAmPMValue` | `toggle-am-p-m-value` | AM/PM toggle value                                     | `string`  | `''`    |
| `value`           | `value`               | Indicate the dropdown input value                      | `string`  | `''`    |


## Slots

| Slot      | Description                          |
| --------- | ------------------------------------ |
| `"input"` | Required slot for input[type="time"] |


## Dependencies

### Used by

 - [gux-form-field](../..)

### Depends on

- [gux-option](../../../gux-dropdown/gux-option)
- [gux-icon](../../../gux-icon)

### Graph
```mermaid
graph TD;
  gux-input-time --> gux-option
  gux-input-time --> gux-icon
  gux-form-field --> gux-input-time
  style gux-input-time fill:#f9f,stroke:#333,stroke-width:4px
```

----------------------------------------------

*Built with [StencilJS](https://stenciljs.com/)*
