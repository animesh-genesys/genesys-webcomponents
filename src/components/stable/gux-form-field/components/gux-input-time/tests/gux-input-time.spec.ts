import { newSpecPage, SpecPage } from '@stencil/core/testing';
import MutationObserver from 'mutation-observer';

import { GuxInputTime } from '../gux-input-time';

const components = [GuxInputTime];
const language = 'en';

describe('gux-input-time', () => {
  let page: SpecPage;

  beforeEach(async () => {
    global.MutationObserver = MutationObserver;

    page = await newSpecPage({
      components,
      html: `
        <gux-input-time>
          <input type="time" slot="input"/>
        </gux-input-time>
      `,
      language
    });
  });

  it('should build', async () => {
    expect(page.rootInstance).toBeInstanceOf(GuxInputTime);
  });

  it('should render', async () => {
    expect(page.root).toMatchSnapshot();
  });
});
