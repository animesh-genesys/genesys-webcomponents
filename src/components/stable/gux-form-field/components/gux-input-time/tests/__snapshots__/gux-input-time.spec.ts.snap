// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`gux-input-time should render 1`] = `
<gux-input-time>
  <input slot="input" type="time">
  <div aria-disabled="false" class="gux-input-time">
    <div class="gux-input-time-field">
      <div class="gux-input-time-field-input">
        <div class="gux-input-time-field-text-input">
          <input aria-label="Input hours" class="gux-input-time-hours" max="12" min="1" type="number" value="12">
          <span class="gux-time-separator">
            :
          </span>
          <input aria-label="Input minutes" class="gux-input-time-minutes" max="59" min="00" type="number" value="00">
          <button aria-label="AM selected, click to toggle" class="gux-input-time-am-pm-selector" type="button">
            AM
          </button>
          <button aria-label="Toggle suggested times" class="gux-clock-button" type="button">
            <gux-icon decorative="" icon-name="clock-outline"></gux-icon>
          </button>
        </div>
      </div>
    </div>
    <div aria-label="time options opened" class="gux-time-options">
      <div class="gux-time-options-list">
        <gux-option class="gux-time-option" value="12:00">
          12:00
        </gux-option>
        <gux-option class="gux-time-option" value="12:15">
          12:15
        </gux-option>
        <gux-option class="gux-time-option" value="12:30">
          12:30
        </gux-option>
        <gux-option class="gux-time-option" value="12:45">
          12:45
        </gux-option>
        <gux-option class="gux-time-option" value="1:00">
          1:00
        </gux-option>
        <gux-option class="gux-time-option" value="1:15">
          1:15
        </gux-option>
        <gux-option class="gux-time-option" value="1:30">
          1:30
        </gux-option>
        <gux-option class="gux-time-option" value="1:45">
          1:45
        </gux-option>
        <gux-option class="gux-time-option" value="2:00">
          2:00
        </gux-option>
        <gux-option class="gux-time-option" value="2:15">
          2:15
        </gux-option>
        <gux-option class="gux-time-option" value="2:30">
          2:30
        </gux-option>
        <gux-option class="gux-time-option" value="2:45">
          2:45
        </gux-option>
        <gux-option class="gux-time-option" value="3:00">
          3:00
        </gux-option>
        <gux-option class="gux-time-option" value="3:15">
          3:15
        </gux-option>
        <gux-option class="gux-time-option" value="3:30">
          3:30
        </gux-option>
        <gux-option class="gux-time-option" value="3:45">
          3:45
        </gux-option>
        <gux-option class="gux-time-option" value="4:00">
          4:00
        </gux-option>
        <gux-option class="gux-time-option" value="4:15">
          4:15
        </gux-option>
        <gux-option class="gux-time-option" value="4:30">
          4:30
        </gux-option>
        <gux-option class="gux-time-option" value="4:45">
          4:45
        </gux-option>
        <gux-option class="gux-time-option" value="5:00">
          5:00
        </gux-option>
        <gux-option class="gux-time-option" value="5:15">
          5:15
        </gux-option>
        <gux-option class="gux-time-option" value="5:30">
          5:30
        </gux-option>
        <gux-option class="gux-time-option" value="5:45">
          5:45
        </gux-option>
        <gux-option class="gux-time-option" value="6:00">
          6:00
        </gux-option>
        <gux-option class="gux-time-option" value="6:15">
          6:15
        </gux-option>
        <gux-option class="gux-time-option" value="6:30">
          6:30
        </gux-option>
        <gux-option class="gux-time-option" value="6:45">
          6:45
        </gux-option>
        <gux-option class="gux-time-option" value="7:00">
          7:00
        </gux-option>
        <gux-option class="gux-time-option" value="7:15">
          7:15
        </gux-option>
        <gux-option class="gux-time-option" value="7:30">
          7:30
        </gux-option>
        <gux-option class="gux-time-option" value="7:45">
          7:45
        </gux-option>
        <gux-option class="gux-time-option" value="8:00">
          8:00
        </gux-option>
        <gux-option class="gux-time-option" value="8:15">
          8:15
        </gux-option>
        <gux-option class="gux-time-option" value="8:30">
          8:30
        </gux-option>
        <gux-option class="gux-time-option" value="8:45">
          8:45
        </gux-option>
        <gux-option class="gux-time-option" value="9:00">
          9:00
        </gux-option>
        <gux-option class="gux-time-option" value="9:15">
          9:15
        </gux-option>
        <gux-option class="gux-time-option" value="9:30">
          9:30
        </gux-option>
        <gux-option class="gux-time-option" value="9:45">
          9:45
        </gux-option>
        <gux-option class="gux-time-option" value="10:00">
          10:00
        </gux-option>
        <gux-option class="gux-time-option" value="10:15">
          10:15
        </gux-option>
        <gux-option class="gux-time-option" value="10:30">
          10:30
        </gux-option>
        <gux-option class="gux-time-option" value="10:45">
          10:45
        </gux-option>
        <gux-option class="gux-time-option" value="11:00">
          11:00
        </gux-option>
        <gux-option class="gux-time-option" value="11:15">
          11:15
        </gux-option>
        <gux-option class="gux-time-option" value="11:30">
          11:30
        </gux-option>
        <gux-option class="gux-time-option" value="11:45">
          11:45
        </gux-option>
      </div>
    </div>
  </div>
</gux-input-time>
`;
